### PR TITLE
Add mdbook-toc, markers, and documentation

### LINKED ISSUE #1014: Add mdbook-toc
[mdbook-toc](https://github.com/badboy/mdbook-toc) is an mdbook preprocessor that allows auto-generating TOCs. This is useful for long sections with many subtitles.

- [x] Add the preprocessor configuration to `book.toml`
- [x] Add the preprocessor to the CI configuration
- [x] Add the preprocessor dependency and usage to the instructions on how to contribute to the guide.
- [x] Identify sections in need of a TOC and add the `<!-- toc -->` marker

### PR DESCRIPTION
**Pull request by igaray**
Closes #1014

- Add the preprocessor configuration to `book.toml`
- Add the preprocessor to the `.travis.yml` CI configuration.
- Add the preprocessor dependency and usage to the instructions on how to contribute to the guide.
- Identify sections in need of a TOC and add the  marker

TBH, it surprised me how useful this is in some sections.

### CODE REVIEW

**Comment by jyn514** [comment]
Can you post a screenshot of some of the TOCs?

**Comment by jyn514** [comment]
Also, it would be nice to split out the pure cleanups into another PR (or just a different commit) so it's easy to tell what changes are relevant.

**Comment by igaray** [comment]
Screenshots:

**Comment by igaray** [comment]
Incoming force-push with the split into separate commits.

**Comment by igaray** [comment]
Split it into three commits, one with the config and CI changes, another adding the toc markers and heading level changes, and the third with whitespace and punctuation fixes.

File: src/contributing.md [hunk 1]
```diff
@@ -1,17 +1,10 @@
 # Contributing to Rust
 
+<!-- toc -->
+
 Thank you for your interest in contributing to Rust! There are many ways to
 contribute, and we appreciate all of them.
```
Comment (jyn514) [inline src/contributing.md:6]: Hmm, I kind of like that the table of contents came after 'there are many ways to contribute', it's like it's showing the ways to contribute. If you move `<!-- toc -->` after the paragraph, will it render the table of contents after too?

**Review by jyn514 — APPROVED** [review]
This looks amazing :tada:

File: src/contributing.md (hunk 1 above)
Comment (igaray) [reply src/contributing.md:6]: Yep! it renders the TOC wherever the marker is, I'll push a change moving it back before merging.

**Pull request merged**